Source/core/FileSystem.cpp

```diff
@@ -56,6 +56,90 @@ namespace Core {
     {
         Close();
     }
+
+    /* static */ string File::Normalize(const string& location, bool& valid)
+    {
+        string result(location);
+
+        valid = true;
+
+        // First see if we are not empy.
+        if (result.empty() == false) {
+            uint32_t index = 0;
+
+           while (index < result.length()) {
+
+#ifdef __WINDOWS__
+                if (result[index] == '\\') {
+                    result[index] = '/';
+                }
+#endif
+
+                if ((result[index] == '/') && (index >= 1) ) {
+                    if (result[index - 1] == '/') {
+                        if (index >= 2) {
+                            // We have a double slash, clear all till the beginning
+                           result = result.substr(index - 1);
+                           index = 1;
+                        }
+                    }
+                    else if (result[index - 1] == '.') {
+                        if ((index == 1) || (result[index - 2] == '/')) {
+                            // It is a dot, remove it...
+                            uint32_t offset = (index == 1 ? 0 : index - 2);
+                            result.erase(offset, 2);
+                            index = offset;
+                        }
+                        else if ((result[index - 2] == '.') && ((index == 2) || (result[index - 3] == '/'))) {
+                            if (index <= 3) {
+                                valid = false;
+                                result.clear();
+                            }
+                            else {
+                                // Seems like we are moving up a directory... execute that on the result... if we can...
+                                // there is data we can drop, drop it, drop it till the '/' is found
+                                uint32_t offset = (index <= 3 ? 0 : index - 4);
```

Comment:
small improvement perhaps for readability: index <= 3 cannot happen here as you already (rightfully) decided above that that would be invalid as you would move out of the allowed path

Reply:
It definitly can happen and it did happen during testing :-) The test I tried:
	bool valid = false;
	string pathName = "../testFile/With..A..Name\\File.txt";
	string outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "\\/B/C/../../testFile/With..A..Name\\File.txt";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "\\/B/C/../../../testFile/With..A..Name\\File.txt";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "/B/C/../../testFile/With..A..Name\\File.txt";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "/B/C/D/../../testFile/../With..A..Name\\File.txt";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "../testFile/../With..A..Name\\File.txt";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "/../testFile/../With..A..Name\\File.txt";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "/..";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "..";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "../";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = ".././testFile/../With..A..Name\\./File.txt";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = ".././././././testFile/../With..A..Name\\./File.txt";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "././././././testFile/../With..A..Name\\./File.txt";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "./../././././testFile/../With..A..Name\\./File.txt";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "/testFile/in/a/directory//With..A..Name\\./File.txt";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "/testFile/in/a/directory/////With..A..Name\\./File.txt\\./././.\\.";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());

	pathName = "/testFile/in/a/directory/////With..A..Name\\./File.txt\\..";
	outcome = Core::File::Normalize(pathName, valid);
	printf("[%s] %s => %s\n", valid ? _T("true") : _T("false"), pathName.c_str(), outcome.c_str());
one of them caused an error which led to the analyses that I had to check for <= 3..